uaclient/data_types.py

```diff
@@ -0,0 +1,133 @@
+from typing import Any, List, Optional, Type, TypeVar
+
+from uaclient.exceptions import UserFacingError
+
+INCORRECT_TYPE_ERROR_MESSAGE = (
+    "Expected value with type {type} but got value: {value}"
+)
+INCORRECT_LIST_ELEMENT_TYPE_ERROR_MESSAGE = (
+    "Got value with incorrect type at index {index}: {nested_msg}"
+)
+INCORRECT_FIELD_TYPE_ERROR_MESSAGE = (
+    'Got value with incorrect type for field "{key}": {nested_msg}'
+)
+
+
+class IncorrectTypeError(UserFacingError):
+    def __init__(self, expected_type: str, got_value: Any):
+        super().__init__(
+            INCORRECT_TYPE_ERROR_MESSAGE.format(
+                type=expected_type, value=repr(got_value)
+            )
+        )
+
+
+class IncorrectListElementTypeError(IncorrectTypeError):
+    def __init__(self, err: IncorrectTypeError, at_index: int):
+        self.msg = INCORRECT_LIST_ELEMENT_TYPE_ERROR_MESSAGE.format(
+            index=at_index, nested_msg=err.msg
+        )
+
+
+class IncorrectFieldTypeError(IncorrectTypeError):
+    def __init__(self, err: IncorrectTypeError, key: str):
+        self.msg = INCORRECT_FIELD_TYPE_ERROR_MESSAGE.format(
+            key=key, nested_msg=err.msg
+        )
+
+
+class DataValue:
+    @staticmethod
+    def from_value(val: Any) -> Any:
+        return val
+
+
+class StringDataValue(DataValue):
+    @staticmethod
+    def from_value(val: Any) -> str:
+        if not isinstance(val, str):
+            raise IncorrectTypeError("string", val)
+        return val
```

Comment:
Can't we have that logic on the `DataValue`  class and just determine the type we expect on the subclasses ?

Reply:
The thing we can't do that way is the return type of the method ` -> str`.
I'd rather keep the extra boilerplate so that we can have the typing.
There might be a fancy way to do this with Generics, but it's nice and simple as it is.